app/javascript/packages/document-capture/components/full-screen.jsx

```diff
@@ -1,81 +1,139 @@
 /** @typedef {import('react').ReactNode} ReactNode */
 
 /**
- * @typedef FullScreenProps
- *
- * @prop {()=>void=} onRequestClose Callback invoked when user initiates close intent.
- * @prop {ReactNode} children       Child elements.
+ * @typedef {()=>void} RequestCloseCallback
  */
 
 /**
- * Number of active instances of FullScreen currently mounted, used in determining when overlay body
- * class should be added or removed.
+ * @typedef FullScreenProps
  *
- * @type {number}
+ * @prop {RequestCloseCallback=} onRequestClose Callback invoked when user initiates close intent.
+ * @prop {string} label Accessible label for modal.
+ * @prop {ReactNode} children Child elements.
  */
-let activeInstances = 0;
 
 /**
- * @param {FullScreenProps} props Props object.
+ * @param {React.MutableRefObject<HTMLElement?>} containerRef
+ * @param {RequestCloseCallback} onRequestClose
  */
-function FullScreen({ onRequestClose = () => {}, children }) {
-  const { t } = useI18n();
-  const { getAssetPath } = useAsset();
-  const trapRef = useRef(/** @type {?import('focus-trap').FocusTrap} */ (null));
+function useFocusTrap(containerRef, onRequestClose) {
+  const trapRef = useRef(/** @type {FocusTrap?} */ (null));
   const onRequestCloseRef = useRef(onRequestClose);
+
   useEffect(() => {
     // Since the focus trap is only initialized once, but the callback could
     // be changed, ensure that the current reference is kept as a mutable value
     // to reference in the deactivation.
     onRequestCloseRef.current = onRequestClose;
   }, [onRequestClose]);
 
-  const setFocusTrapRef = useCallback((node) => {
-    if (trapRef.current) {
-      trapRef.current.deactivate();
-    }
-
-    if (node) {
-      trapRef.current = createFocusTrap(node, {
+  useEffect(() => {
+    if (containerRef.current) {
+      trapRef.current = createFocusTrap(containerRef.current, {
         onDeactivate: () => onRequestCloseRef.current(),
         clickOutsideDeactivates: true,
       });
 
       trapRef.current.activate();
     }
-  }, []);
-
-  useEffect(() => {
-    if (activeInstances++ === 0) {
-      document.body.classList.add('has-full-screen-overlay');
-      document.documentElement.classList.add('has-full-screen-overlay');
-    }
 
     return () => {
-      if (--activeInstances === 0) {
-        document.body.classList.remove('has-full-screen-overlay');
-        document.documentElement.classList.remove('has-full-screen-overlay');
-      }
+      trapRef.current?.deactivate();
     };
   }, []);
+}
+
+const useBodyClass = (() => {
+  /**
+   * @type {WeakMap<any, number>}
+   */
+  const activeInstancesByComponent = new WeakMap();
 
   return (
-    <div ref={setFocusTrapRef} aria-modal="true" className="full-screen bg-white">
+    /**
+     * @param {string} className Class name to add to body element
+     * @param {React.ComponentType<any>} Component React component definition
+     */
+    (className, Component) => {
```

Comment:
how come `Component` is capitalized here? it's still a variable right?

Reply:
> how come `Component` is capitalized here? it's still a variable right?

I capitalized it because it's referring to a React component type, which can be meaningful at least in the context of using it for JSX:

>Capitalized types indicate that the JSX tag is referring to a React component.

https://reactjs.org/docs/jsx-in-depth.html#specifying-the-react-element-type

This is somewhat unique in that it's being passed around as an argument, and we're not using it for JSX rendering.

Since it's a React-ism, I poked around React's codebase to see how they name it. I think the usage here is consistent.

Examples:

- https://github.com/facebook/react/blob/e0d9b28/packages/react-debug-tools/src/ReactDebugHooks.js#L654
- https://github.com/facebook/react/blob/e0d9b28/packages/react-dom/src/server/ReactPartialRenderer.js#L283